### ISSUE TITLE
Developers targeting browser-wasm can use Web Crypto APIs

### ISSUE THREAD

**Issue by marek-safar** [dotnet/runtime#40074]
We want to avoid shipping OpenSSL for Browser as that’s not something that aligns well with web nature of WebAssembly as well as it has noticeable impacts on the size of the final app. This would also save us from having to deal with zero-day vulnerabilities as well as support for crypto algorithms which are only secure if they use special CPU instructions (RNG and AES for example).

Using the platform native crypto functions is the preferred solution as it does not have any noticeable size and it’s also the most performant solution. All browser have nowadays support for Crypto APIs which we should be able to use to implement the core crypto functions required by BCL. 

Relevant documentation can be found at https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto. The tricky part will be to deal with the async nature of these APIs but we could introduce new async APIs to make the integration easier.

.NET has support for old/obscure algorithms and also other features like certificates which are not relevant in browser space and for them, we would keep throwing PNSE.

### Design Proposal

The design proposal for how we will enable cryptographic algorithms in .NET for WebAssembly can be found here:
https://github.com/dotnet/designs/blob/main/accepted/2021/blazor-wasm-crypto.md

We intend on providing the most secure implementations of cryptographic algorithms when available. On browsers which support `SharedArrayBuffer`, we will utilize that as a synchronization mechanism to perform a sync-over-async operation, letting the browser's secure `SubtleCrypto` implementation act as our cryptographic primitive. On browsers which do not support `SharedArrayBuffer`, we will fall back to in-box managed algorithm implementations of these primitives.

Where managed algorithm implementations are needed, they can be migrated from the .NET Framework Reference Source.

* https://github.com/microsoft/referencesource/tree/master/mscorlib/system/security/cryptography
* https://github.com/microsoft/referencesource/tree/master/System.Core/System/Security/Cryptography

### Work Items

- [x] SHA1
- [x] SHA256
- [x] SHA384
- [x] SHA512
- [x] HMACSHA1
- [x] HMACSHA256
- [x] HMACSHA384
- [x] HMACSHA512
- [x] AES-CBC
  - [x] AES-ECB and AES-CFB should throw PlatformNotSupportedException, as they are not available in webcrypto
- [x] PBKDF2 (via [Rfc2898DeriveBytes](https://docs.microsoft.com/dotnet/api/system.security.cryptography.rfc2898derivebytes)) 
- [x] HKDF

### Acceptance Criteria

- [ ] Integrate into arch-wasm infrastructure
- [ ] Ensure a productive development/testing workflow with WebAssembly
- [ ] Get all existing unit tests for the algorithms passing in WebAssembly
- [ ] Look for opportunities to improve argument validation
- [ ] Look for opportunities to modernize the code, such as using Span
- [ ] Validate end-to-end scenarios in desktop browsers that support `SharedArrayBuffer`
- [ ] Validate end-to-end scenarios in mobile browsers that support `SharedArrayBuffer`
- [ ] Validate end-to-end scenarios in desktop browsers that do not support `SharedArrayBuffer`
- [ ] Validate end-to-end scenarios in mobile browsers that do not support `SharedArrayBuffer`

### Issues to address 
- [ ] https://github.com/dotnet/runtime/issues/69741
- [x] https://github.com/dotnet/runtime/issues/69740
- [x] https://github.com/dotnet/runtime/issues/69806

**Comment by jaykrell** [comment]
> only secure if they use special CPU instructions (RNG and AES for example

Really "only secure"? Or just faster?

**Comment by bartonjs** [comment]
> Really "only secure"? Or just faster?

AES is literally only secure with dedicated hardware instructions on modern CPUs.  All software implementations are subject to various timing problems that ultimately leak the key to a dedicated attack (http://cr.yp.to/antiforgery/cachetiming-20050414.pdf).

**Comment by GrabYourPitchforks** [comment]
As part of this work item, we should also verify that `RandomNumberGenerator` is correctly plumbed down to [`crypto.getRandomValues`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues). Right now it's plumbed through `/dev/random`, which I think the Mono wasm engine ties through correctly, but it's always good to remove indirection and points of failure where possible.

**Comment by kjpou1** [comment]
Looking into support and providing a prototype that works with browser os.

**Comment by juanfranblanco** [comment]
@marek-safar Do we have a list of all the crypto libraries that are not supported and the best way to identify if we are running on wasm to fallback to other libraries?

**Comment by marek-safar** [comment]
@juanfranblanco the best way to verify a library is fully wasm compatible is to enable platform compatibility analyzer for `browser-wasm`. More details at https://docs.microsoft.com/en-us/dotnet/standard/analyzers/platform-compat-analyzer

**Comment by NicolasDorier** [comment]
Some users of my library (NBitcoin) are unable to use it because of crypto primitives not implemented.

However, looking at the source code, it has already been implemented on managed side: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/SHAHashProvider.Browser.cs

I think I will just copy/paste this code and ship it inside my library instead of relying on the dotnet's implementation.
The alternatives (using Bouncy Castle, which is too slow) or not supporting my library on WASM is not possible.
I can't even do some conditional pre compilation magic with `#if` as you can't create a library targetting wasm specifically...

However, before I continue, I would like to know: Why did you implemented it, and turned it off?

**Comment by marek-safar** [comment]
Which crypto support are you missing? The latest .NET5 release includes support for core primitive types and ships with the code you referenced.

**Comment by agnauck** [comment]
@NicolasDorier instead of `#if` defines you can either catch the `PlatformNotSupportedException` or use the RuntimeInformation object to detect whether you code is running on WASM Browser or not.

I included all the missing codes in my library as well until they ship in NET6 for WASM as well.
See eg here:
https://github.com/matrix-xmpp/matrix-vnext/blob/75d3951d26fc4539b974bedfac41c2fe4a551631/src/Matrix.Core/Crypt/PBKDF2.cs#L15
https://github.com/matrix-xmpp/matrix-vnext/tree/net5/src/Matrix.Core/Crypt/Wasm

**Comment by GrabYourPitchforks** [comment]
For anybody seeing failures from the crypto stack on wasm, please include details of the exception you're seeing, including the exact error text and stack trace. The SHA\* functions should behave correctly in wasm / browser. (Note: this _does not_ include HMACSHA\*, at least for .NET 5.) If folks are seeing errors we'll need the exception information to help track down the failure and to provide a workaround. Thanks!

**Comment by NicolasDorier** [comment]
@agnauck catching `PlatformNotSupportedException` is not possible, our GetHash is very often in a hot path so try/catch would have user noticeable impact.

@marek-safar I thought the user complaining about it was using on .NET5.0. I will investigate. We have a dependency on `HMACSHA` as well, but this is another matter.

**Comment by agnauck** [comment]
@NicolasDorier then just check if you are running on Wasm platform.

**Comment by arivera12** [comment]
I had `RijndaelManaged` working properly on Blazor WAsm 3.1. Now that I upgraded to NET5 is broken!

![image](https://user-images.githubusercontent.com/5600102/108652425-84661400-749a-11eb-83e9-43963b2b3a17.png)

**Comment by arivera12** [comment]
For anyone hitting this error for blazor wasm and encryption is needed in all ways I used https://github.com/sytelus/CryptoJS  in the meantime and works perfectly.

**Comment by DemiMarie** [comment]
> > Really "only secure"? Or just faster?
> 
> AES is literally only secure with dedicated hardware instructions on modern CPUs. All software implementations are subject to various timing problems that ultimately leak the key to a dedicated attack (http://cr.yp.to/antiforgery/cachetiming-20050414.pdf).

Nit: It is possible to implement AES securely without any special instructions.  However, the resulting implementation is extremely slow, which is why nobody does it.

**Comment by GrabYourPitchforks** [comment]
> It is possible to implement AES securely without any special instructions.

Only if you can write raw assembly. This precludes the ability to write a secure implementation in managed or wasm code. That's one of the reasons we don't want to ship our own implementation separate from the browser / OS-provided implementation.

**Comment by arivera12** [comment]
It's this gonna be fixed on net 6? 

I even notice that signal r is broken on .net 5....

 I downgraded to net 3.1 for blazor wasm.

**Comment by DemiMarie** [comment]
> > It is possible to implement AES securely without any special instructions.
> 
> Only if you can write raw assembly. This precludes the ability to write a secure implementation in managed or wasm code. That's one of the reasons we don't want to ship our own implementation separate from the browser / OS-provided implementation.

You don’t need to write raw assembly to do it.  See BearSSL for an example of a secure (although very slow) implementation in pure portable C.

**Comment by andyhopp** [comment]
Do we have any indication whether .NET 6 will ship with support for cryptographic algorithms? I just tried the latest preview (6.0.100-preview.4.21255.9 as of this comment) and still get the dreaded "System.Security.Cryptography.Algorithms is not supported on this platform." exception.

**Comment by DemiMarie** [comment]
> Do we have any indication whether .NET 6 will ship with support for cryptographic algorithms? I just tried the latest preview (6.0.100-preview.4.21255.9 as of this comment) and still get the dreaded "System.Security.Cryptography.Algorithms is not supported on this platform." exception.

The main problem, as far as I can tell, is that the Web Crypto API is inherently asynchronous, while the .NET cryptographic APIs are synchronous.  That means that using the Web Crypto API to implement the .NET cryptographic APIs is not possible.  This is arguably a design flaw in the Web Crypto API.

From what I can tell, your best option for now is to call Web Crypto API yourself.

**Comment by Xyncgas** [comment]
new people keep coming in or being redirected from duplicate issues about this problem, they are trying to explain their needs and why it is important, but can you folks please update whether you are planning to control web crypto from wasmland or are you planning to actually re-enable blazor to be able to use system.cryptography again without all the draw backs or what is your plan?
How long are we going to wait? Is it something you are figuring out rn or are you guys just waiting to see alternatives? Is it short term or long term?
Is microsoft going to provide alternatives because crytography is essential (working on a nuget package that wraps it through js interop)? Or are we on our own for now no matter how important it is that we are saying it is in this issue? Can I please get a signal?

**Comment by Xyncgas** [comment]
> > Do we have any indication whether .NET 6 will ship with support for cryptographic algorithms? I just tried the latest preview (6.0.100-preview.4.21255.9 as of this comment) and still get the dreaded "System.Security.Cryptography.Algorithms is not supported on this platform." exception.
> 
> The main problem, as far as I can tell, is that the Web Crypto API is inherently asynchronous, while the .NET cryptographic APIs are synchronous. That means that using the Web Crypto API to implement the .NET cryptographic APIs is not possible. This is arguably a design flaw in the Web Crypto API.
> 
> From what I can tell, your best option for now is to call Web Crypto API yourself.

why can't we just solve everything using c / c++

**Comment by arivera12** [comment]
This issue is quite annoying and lot of people have asked whether this gonna be resolve and no response at all. 

As I mentioned on a top comment the only working version of `System.Security.Cryptography.Algorithms` is blazor wasm using Net 3.1 otherwise upgrading will throw this error, and any library related using this will throw the same error. 

I even mentioned that `Signal R` is broken because is using the `System.Security.Cryptography.Algorithms` under the hood and makes unusable over blazor wasm.

Soon we gonna have a year on july and no response.

**Comment by GrabYourPitchforks** [comment]
For folks experiencing problems with SignalR client - can you please provide the full exception info, including error message & full stack trace? The full stack trace will help us determine which specific component is triggering this error.

**Comment by andyhopp** [comment]
> For folks experiencing problems with SignalR client - can you please provide the full exception info, including error message & full stack trace? The full stack trace will help us determine which specific component is triggering this error.

First: I presume you're asking this so you know which classes to patch. I apologize in advance if that wasn't your intent. Now...

Not to be a jerk (he said, preparing to be a jerk...), but patching downstream libraries to invoke the low-level API is not the right solution; it provides an opportunity for someone to "do it wrong." Fixing it in the runtime fixes it for everyone.

**Comment by GrabYourPitchforks** [comment]
> First: I presume you're asking this so you know which classes to patch. I apologize in advance if that wasn't your intent.

There are lots of options we're looking at, including shimming the existing APIs, creating new APIs, or patching targeted call sites. The ultimate course of action will depend on what options we believe we can provide securely while still fulfilling any legal or regulatory obligations.

That said, if we don't have the information like stack traces needed to investigate, then the current situation remains and _nobody_ benefits. Not people who are calling the APIs directly. Not people who are calling the APIs indirectly through SignalR. No one.

**Comment by igandrews** [comment]
I saw mention above that SHA512 should work with .net 5 but I'm getting the following: 
![image](https://user-images.githubusercontent.com/17983788/121423803-a5b21980-c93e-11eb-8a70-d43f3ba7cfea.png)

**Comment by GrabYourPitchforks** [comment]
@igandrews The packages _System.Security.Cryptography.Cng_ and _System.Security.Cryptography.Csp_ contain Windows-specific APIs. There's a minimum amount of shimming for Linux, but otherwise most APIs in those packages will fail on all non-Windows platforms. (You'll see the same failures on iOS, Android, etc. as you would in wasm.)

Consider using `SHA512.Create` instead of `new SHA512CryptoServiceProvider`. This pattern should work across all runtimes: Windows, Linux, iOS, Android, wasm, and others.

**Comment by igandrews** [comment]
Great! Thx. Will check that out. Hope I didn't inadvertently hijack this thread but thought since I'm seeing a PlatformNotSupportException around the crypto apis only in wasm that this was related but maybe this will help someone else. Note I'm not _directly_ referencing that package so wondering how I would have known to look for that and that such api is using/requires platform specific apis and what alternative I might have evaluated?

**Comment by GrabYourPitchforks** [comment]
@igandrews There has been a drive starting with 5.0 and continuing into 6.0 to annotate APIs that are known to work only on specific platforms. The general idea is that if you're creating an app that targets OS-A but you consume an API that works only on OS-B, you'll receive a compile-time warning.

In the case of _System.Security.Cryptography.Cng_ and _System.Security.Cryptography.Csp_, we've started obsoleting these types. `SHA512CryptoServiceProvider` as of .NET 6 Preview 5 (which should be available in a few weeks) is now obsolete as warning with a message directing callers to use `SHA512.Create` instead. (See https://github.com/dotnet/runtime/issues/46934.) So this should get better with the coming releases.

**Comment by Xyncgas** [comment]
Once again, anytime you are releasing a new version with new features and obsolete features, just make obsolete features optional and if there is conflict then let user choose, LET USERS CHOOSE, don't force them, I think it just solves everything, but no one listening though.
So At Least People's Code Can Still Build, Compile, Run.

**Comment by GrabYourPitchforks** [comment]
We just posted our plan for cryptographic support in Blazor/wasm apps in .NET 7 and beyond. [See here](https://github.com/dotnet/designs/blob/main/accepted/2021/blazor-wasm-crypto.md) for full details. That document also includes sample code demonstrating how to interop with the JS layer to get support for other algorithms (like HMACSHA256) which aren't supported out-of-the-box in Blazor/wasm in .NET 6.

Please use this issue (https://github.com/dotnet/runtime/issues/40074) for feedback on that document, including proposing scenarios for consideration.

Thanks!

**Comment by akoeplinger** [comment]
Small comment about this section:
> Organizations which require FIPS 140-2 compliant implementations or implementations which make stronger side channel guarantees must ensure that their Blazor client code executes only in browsers which support SharedArrayBuffer and which carry FIPS 140-2 compliant primitives via SubtleCrypto.

We could add an AppContext switch to turn off the managed fallback so users can enforce they're not using the fallback.

**Comment by vcsjones** [comment]
> We could add an AppContext switch to turn off the managed fallback so users can enforce they're not using the fallback.

All of the FIPS checks were removed from .NET Core and I personally would not like the return of options to disable things to support FIPS requirements. In the case of OpenSSL it won't be FIPS either unless you are using a FIPS build of OpenSSL, which most distros do not use.

**Comment by DemiMarie** [comment]
I strongly recommend _not_ supporting a managed AES implementation by default; it will either be super slow or vulnerable to side-channel attacks.  ChaCha20-Poly1305, on the other hand, can be reasonably fast in a pure C implementation, which I recommend using.

**Comment by akoeplinger** [comment]
@vcsjones yeah, though it could be useful to make sure the managed implementation is trimmed out for size reasons regardless of FIPS issues.

**Comment by GrabYourPitchforks** [comment]
> We could add an AppContext switch to turn off the managed fallback so users can enforce they're not using the fallback.

We brought this up with the Crypto Board, because our original plan was to implement exactly this. The strongly advised against doing this. Our AppContext switch would be seen as a "do you want a compliant algorithm or not?" setting. But since we can't make any guarantees about the browser's underlying implementation, we don't know anything about compliance, so the switch doesn't actually answer the question people want it to answer. It would be confusing at best, misleading at worst.

> I strongly recommend not supporting a managed AES implementation by default; it will either be super slow or vulnerable to side-channel attacks. ChaCha20-Poly1305, on the other hand, can be reasonably fast in a pure C implementation, which I recommend using.

Per the doc, performance and side-channel protection is an explicit non-goal of our fallback implementations. We are not implementing ChaCha20-Poly1305.

**Comment by gulbanana** [comment]
Feedback: I have a complex internal codebase, originally written for .NET Framework, later ported to .NET Core, now being used in Blazor applications. We use these features of System.Security.Cryptography:
* ECDSA for signing some content
* RSA or PBKDF2+RSA for authentication
* AES for private communication channels 

We first used Blazor in .NET Core 3.1 We disabled the elliptic-curve related features (which are not core to the application) and the rest all works in WebAssembly, presumably using the managed implementations from the mono runtime. In .NET 5, keyed encryption managed implementations were removed and we weren't able to update. Instead, we've been using Blazor Server, which meets our use cases for now but perhaps not forever.

These applications are actively maintained and we don't mind adding async code paths - the virality is real, but it's a known quantity and most encryption takes place on a code path that already contains some form of I/O (most of which we have made async already). Until now, I was hoping that .NET 6 would either include new async APIs or restore the managed-code implementations (if this could somehow be made secure).

Given the plan you've set out, here are my organisation's answers to the key questions :

> 1. Is it acceptable for the client to have this key in the first place? Would it present a risk to your service if the client hooked a debugger to the wasm app and dumped the raw key material?

Yes, it's acceptable. We have two kinds of client side keys: asymmetric keys derived from a user-entered password, and symmetric keys which have been provided by the server for a session, secured by the other ones or by out-of-band mechanisms (oauth etc). These are not static secrets and would be ok for a user to extract.

> 2. Is it acceptable for the client to fall back to a managed algorithm which may not be free of side channels? If not, consider whether it would be safer to perform your cryptography entirely within JavaScript using the Promise-based SubtleCrypto API, having your application make an async-based call into your JavaScript-based helper.

Fallback with reduced security guarantees worries us. That's effectively what we had in 3.1, but it's probably not a risk we should have been accepting. We will probably do as you suggest, bite the bullet and write SubtleCrypto-based client-side encryption. In the past, I've had the impression that Blazor JS interop is 'fragile' or somehow not to be relied upon for inner-loop functionality like this, but that assumption needs reexamining. After using Blazor Server successfully for a while (which is fundamentally interop-based) it does seem to be pretty reliable.

Writing this helper code is doable but it doesn't feel great. Cryptographic code is the sort of thing that you want done once, by experts - by going down this path we will be increasing the degree to which we "roll our own". At least SubtleCrypto seems to be carefully designed to expose APIs that are difficult to misuse.

**Comment by gulbanana** [comment]
There will also be a significant release-engineering cost to us. We'll have to figure out how to do things like alter our nuget packages to include a RID-specific(?) browser build of a dll. This approach implies some sort of support matrix like 

```
+----------------+---------+----------------+
| TFM            | RID(?)  | Implementation |
+----------------+---------+----------------+
| netstandard2.0 | any     | S.S.C.         |
| net5.0         | any     | S.S.C.         |
| net5.0         | browser | SubtleCrypto   |
+----------------+---------+----------------+
```

...which, again, feels like something that would be more within a platform vendor's capabilities than ours as application developers :(

**Comment by DemiMarie** [comment]
> > I strongly recommend not supporting a managed AES implementation by default; it will either be super slow or vulnerable to side-channel attacks. ChaCha20-Poly1305, on the other hand, can be reasonably fast in a pure C implementation, which I recommend using.
> 
> Per the doc, performance and side-channel protection is an explicit non-goal of our fallback implementations. We are not implementing ChaCha20-Poly1305.

Having the fallback be slow is fine, but having it be vulnerable to cache-based timing attacks is not.

**Comment by vcsjones** [comment]
> but having it be vulnerable to cache-based timing attacks is not.

I would like to better understand your concern about timing attacks in a browser scenario. Browser sandboxing has implemented numerous changes to prevent timing attacks in a browser scenario, including site-isolations, and removing high precision timers<sup>1, 2</sup> such as `SharedArrayBuffer` or `performance.now` either by disabling them entirely or reducing their accuracy. Browsers that do implement those APIs require COOP and COEP<sup>3</sup> which tighten the sandboxing even further, but if you have access to `SharedArrayBuffer` then you should end up with the WebCrypto getting used, anyway.

<sup>1</sup>: https://webkit.org/blog/8048/what-spectre-and-meltdown-mean-for-webkit/
<sup>2</sup>: https://developer.mozilla.org/en-US/docs/Web/API/Performance/now#reduced_time_precision
<sup>3</sup>: https://web.dev/coop-coep/

**Comment by GrabYourPitchforks** [comment]
> having it be vulnerable to cache-based timing attacks is not.

Don't get me wrong - we're not going to be sloppy with the implementation. We will make a reasonable effort to balance maintainability / readability of code and security. But we're not going to go out of our way to try to flush caches, to mathematically prove that all code paths take the same cycle count in the face of CPU micro-opts, etc. This is what I meant by us not making guarantees about the implementation being free of side-channels.

The browsers' implementations of SubtleCrypto APIs _hopefully_ provide better defenses against these than we can (but we can't make guarantees about their implementations!), and the linked doc explains both: (a) the scenarios in which we will call them on your behalf rather than relying on our fallbacks; and (b) how to invoke them manually if you want to force their usage over the fallbacks' usage.

**Comment by DemiMarie** [comment]
> > having it be vulnerable to cache-based timing attacks is not.
> 
> Don't get me wrong - we're not going to be sloppy with the implementation. We will make a reasonable effort to balance maintainability / readability of code and security. But we're not going to go out of our way to try to flush caches, to mathematically prove that all code paths take the same cycle count in the face of CPU micro-opts, etc. This is what I meant by us not making guarantees about the implementation being free of side-channels.

Ah okay.  You should probably explain that in the document; my first thought was a naive table-based AES implementation!

> The browsers' implementations of SubtleCrypto APIs _hopefully_ provide better defenses against these than we can (but we can't make guarantees about their implementations!), and the linked doc explains both: (a) the scenarios in which we will call them on your behalf rather than relying on our fallbacks; and (b) how to invoke them manually if you want to force their usage over the fallbacks' usage.

A way to check if the fallback is in use would be nice.

**Comment by bartonjs** [comment]
> We'll have to figure out how to do things like alter our nuget packages to include a browser-specific build of a dll

@gulbanana If you want your libraries to be as small as possible you'd need to make the split as you described, but if you're OK carrying around the browser interop types in your non-browser DLL (which is probably only a couple of KB) you can just use `System.OperatingSystem.IsBrowser()` to make a runtime decision

**Comment by gulbanana** [comment]
That should work in some cases, but are we talking about referencing Microsoft.AspnetCore.Components and expecting all consumers to pass in an IJSRuntime? These libraries are also used by non-Blazor code.
I guess even new async APIs would have the same problem as they wouldn't be added to older TFMs...

**Comment by Theoistic** [comment]
as x509Certificate2 and such is a partial class, you could just add support for System.Security.Cryptographic.Wasm or something which should override the method calls. I had to write an entire subtleCrypto wasm proxy to get my application working. and casting the browser CryptoKey to .NET was a living nightmare.. its a temp fix as i only need to verify ECDSA, RSA-PSS signatures but i know, within 5 days i'll run in to another crypto class not being .NET Supported on wasm.

**Comment by SteveSandersonMS** [comment]
> On browsers which support SharedArrayBuffer, we will utilize that as a synchronization mechanism to perform a sync-over-async operation

Could somebody clarify how this will work? Is there a prototype somewhere?

It sounds like we intend to make the browser's main thread block while an async task occurs in a background thread. But I was under the impression that browsers don't allow the main thread to be blocked, and only background threads can block.

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/wait

**Comment by SteveSandersonMS** [comment]
> Once the main thread has indicated its request it will spin, waiting on a value in the communication buffer.

I thought this was the fatal flaw in the previous prototype? We can't spin on the main thread without risking the app getting killed by the browser.

I understand that for encrypting sufficiently small amounts of data, the process will likely complete in much less than 1s and therefore will usually succeed. But are we OK that with larger larger amounts of data, the app may just crash?

**Comment by Genbox** [comment]
> Developers who are in the uncommon situation of needing to encrypt/decrypt/hash a very large volume of data client-side will have to reach the underlying JS APIs

There are a couple of issues with that:
1. "large" is a relative term that is not measured in bytes - but in time. Slow/overburdened CPUs, inefficient implementations (both in the web crypto APIs and user code), and larger encryption payloads can all contribute to the issue.
2. The whole reason behind the .NET Blazor layer is to avoid the interop. Not saying that Microsoft should solve all interop issues at all costs, but I am saying that as long as crypto is inefficient/unusable Blazer will never be taken seriously.
3. Asking the user to take control over the process might create 'leaking' issues. It is the same idea behind "don't create your own crypto" - do it once and do it right, don't leave the user with responsibilities they can't correctly accomplish without creating holes in their own security.

If there are no other options on the table, I'd go with new async crypto APIs instead. That is just my 2 cents as someone who needs to hash/encrypt large files in a streaming fashion through Blazor but currently can't due to the current limitations.

**Comment by robertmclaws** [comment]
Recommend changing the name of this proposal to "Cryptography APIs", as there are also Web3 APIs that allow browsers to communicate with extensions like MetaMask and TrustWallet. I was hoping that's what this proposal was for.

**Comment by bartonjs** [comment]
@robertmclaws MetaMask and TrustWallet are both about crypto-currency.  This issue is about cryptography itself.  Things like SHA-2-256.  On the browser that's provided by the [Web Crypto API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API)

**Comment by robertmclaws** [comment]
Yep, totally understand.

Wanted to point out that the lack of support for these APIs is presently blocking us from using WalletConnectSharp and Nethereum natively in a Blazor app. We will have to fall back on JSInterop for this, but it is not ideal.

**Comment by egil** [comment]
I just want to inject a small wish for whatever API will surface this in C#: Please make sure it's mockable/replaceable during unit testing (bUnit), and ideally not directly through `IJSRuntime`. 

Hopefully a mockable implementation will be used that uses `IJSRuntime` under the hood, but is exposed under its own interface, e.g. `IBrowserCrypto`, or an abstract base type, like `NavigationManager`. Interface preferred.

Ps. I only scanned through the many comments here, so apologies if this comment is redundant.

**Comment by egil** [comment]
> @egil The implementation doesn't use `IJSRuntime`. It's just a baked-in implementation of some `System.Security.Cryptography` APIs that happens to use browser features as an internal implementation detail.
> 
> AFAIK there would be no need to mock or replace this in a test, since in a test running on CoreCLR the same APIs would work using the native CoreCLR implementations. It's equivalent to any other System.* API.
> 
> It's like using System.TimeZoneInfo - you wouldn't mock it, even though CoreCLR and dotnet.wasm have different implementations which should behave indistinguishably.

Perfect. I just saw some sample code in a linked issue that seemed to use Jsinterop so I got a little worried.

**Comment by Eagle3386** [comment]
Since nobody mentioned it thus far, I'd like to request addition or even incorporation of [Argon2](/mheyman/Isopoh.Cryptography.Argon2) as it's not only written in (quote) `100% managed-code`, but mostly because Argon2 surpasses PBKDF2 _way_ too much regarding security to not be added.

Furthermore, the aforementioned GH repo's author explicitly [states](/mheyman/Isopoh.Cryptography.Argon2#production-status--support)
> You should be aware that this project is supported solely by me and provided as is.

... which, at least to me, translates to something like `if things go totally wrong, bugs or exploits might not be fixed` - which I assume to not be the case for an official .NET implementation.

**Comment by jeffhandley** [comment]
We've concluded that using the `SharedArrayBuffer` approach via a worker is not yielding the end-to-end experience or quality necessary to release this integration. As a result, we've changed the scope of what will be included in .NET 7 to no longer include AES-CBC. Additionally, instead of using SubtleCrypto via the `SharedArrayBuffer` to reach the native implementations of other algorithms, we will rely wholly on managed implementations.

Please see https://github.com/dotnet/runtime/issues/73858 for more information and refer to the [design document](https://github.com/dotnet/designs/blob/main/accepted/2021/blazor-wasm-crypto.md#discussion-1) for more information about the managed implementation expectations.

Where native implementations need to be accessed, they will need to be called via JavaScript interop.
